Torchinductor randn_like lowering

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #93005
* #93004
* #93003

Add lowering for randn_like, fixes https://github.com/pytorch/pytorch/issues/92368 by virtue of not taking a fallback path, although the 0-element prim stride is still incorrect. Would be nice to submit as a decomposition, but that is blocked by https://github.com/pytorch/pytorch/issues/92920. 


cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

eellison committed:
Randn like

[ghstack-poisoned]

eellison committed:
Update on "Torchinductor randn_like lowering"


Add lowering for randn_like, fixes https://github.com/pytorch/pytorch/issues/92368 by virtue of not taking a fallback path, although the 0-element prim stride is still incorrect. Would be nice to submit as a decomposition, but that is blocked by https://github.com/pytorch/pytorch/issues/92920. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Torchinductor randn_like lowering"


Add lowering for randn_like, fixes https://github.com/pytorch/pytorch/issues/92368 by virtue of not taking a fallback path, although the 0-element prim stride is still incorrect. Would be nice to submit as a decomposition, but that is blocked by https://github.com/pytorch/pytorch/issues/92920. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ngimel commented on torch/_inductor/lowering.py:
```diff
@@ -1116,6 +1119,20 @@ def randn(*args, **kwargs):
+        device = kwargs.pop("device", None)
+        device = x.get_device() if device is None else device
+
+        return fast_randn(x.get_size(), dtype=dtype, device=device, **kwargs)
```
hm, this won't preserve permutation if `x` is permuted?

eellison replied:
it won't, but the MO thus far is we only set permutation requirements on graph outputs

eellison committed:
Update on "Torchinductor randn_like lowering"


Add lowering for randn_like, fixes https://github.com/pytorch/pytorch/issues/92368 by virtue of not taking a fallback path, although the 0-element prim stride is still incorrect. Would be nice to submit as a decomposition, but that is blocked by https://github.com/pytorch/pytorch/issues/92920. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ngimel reviewed (approved):
Can you add `rand_like` lowering while you are at it?

eellison committed:
Update on "Torchinductor randn_like lowering"


Add lowering for randn_like, fixes https://github.com/pytorch/pytorch/issues/92368 by virtue of not taking a fallback path, although the 0-element prim stride is still incorrect. Would be nice to submit as a decomposition, but that is blocked by https://github.com/pytorch/pytorch/issues/92920. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Torchinductor randn_like lowering"


Add lowering for randn_like, fixes https://github.com/pytorch/pytorch/issues/92368 by virtue of not taking a fallback path, although the 0-element prim stride is still incorrect. Would be nice to submit as a decomposition, but that is blocked by https://github.com/pytorch/pytorch/issues/92920. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Torchinductor randn_like lowering"


Add lowering for randn_like, fixes https://github.com/pytorch/pytorch/issues/92368 by virtue of not taking a fallback path, although the 0-element prim stride is still incorrect. Would be nice to submit as a decomposition, but that is blocked by https://github.com/pytorch/pytorch/issues/92920. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ngimel commented on torch/_inductor/lowering.py:
```diff
@@ -1643,6 +1643,7 @@ def constant_like(fill_value):
 ones_like = create_tensor_like(tensor_constructor(1))
 if not config.fallback_random:
     rand_like = register_lowering(aten.rand_like)(create_tensor_like(rand))
+    randn_like = register_lowering(aten.randn_like)(create_tensor_like(randn))
```
lol nices, everything existed already

eellison committed:
Update on "Torchinductor randn_like lowering"


Add lowering for randn_like, fixes https://github.com/pytorch/pytorch/issues/92368 by virtue of not taking a fallback path, although the 0-element prim stride is still incorrect. Would be nice to submit as a decomposition, but that is blocked by https://github.com/pytorch/pytorch/issues/92920. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison commented:
@pytorchbot merge -f "unrelated failure"
